File: server/internal/dlp/dlp.go
// Package dlp implements content classification and policy enforcement for Data Loss
// Prevention. It scans text for sensitive-data patterns (PII, PCI, secrets, source code),
// validates candidates (e.g. Luhn for cards, entropy for secrets), and renders a policy
// verdict (audit / alert / block) per channel.
package dlp

import (
	"math"
	"regexp"
	"strings"

	"github.com/sentinel/server/internal/model"
)

// Classifier identifies one class of sensitive data.
type Classifier struct {
	Name     string
	Label    string
	Severity model.Severity
	re       *regexp.Regexp
	validate func(string) bool
}

// Finding is a single classifier hit.
type Finding struct {
	Classifier string
	Label      string
	Severity   model.Severity
	Matches    int
	Sample     string
}

// Policy maps (classifier, channel) to a verdict.
type Policy struct {
	Classifier string // "*" = any
	Channel    string // usb|scp|rsync|ftp|http_upload|email|cloud|"*"
	Verdict    string // audit|alert|block
}

// Engine holds classifiers and policies.
type Engine struct {
	classifiers []Classifier
	policies    []Policy
}

// New builds the default classifier + policy set.
func New() *Engine {
	e := &Engine{}
	e.classifiers = []Classifier{
		{Name: "pci_card", Label: "Payment card number", Severity: model.SevHigh,
			re: regexp.MustCompile(`\b(?:\d[ -]?){13,19}\b`), validate: luhnValid},
		{Name: "pii_ssn", Label: "US Social Security Number", Severity: model.SevHigh,
			re: regexp.MustCompile(`\b\d{3}-\d{2}-\d{4}\b`)},
		{Name: "pii_email", Label: "Email address (bulk)", Severity: model.SevLow,
			re: regexp.MustCompile(`\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b`)},
		{Name: "secret_aws", Label: "AWS access key", Severity: model.SevCritical,
			re: regexp.MustCompile(`\b(?:AKIA|ASIA)[0-9A-Z]{16}\b`)},
		{Name: "secret_privkey", Label: "Private key material", Severity: model.SevCritical,
			re: regexp.MustCompile(`-----BEGIN (?:RSA |EC |OPENSSH |DSA |PGP )?PRIVATE KEY-----`)},
		{Name: "secret_token", Label: "High-entropy API token", Severity: model.SevHigh,
			re: regexp.MustCompile(`\b[A-Za-z0-9_\-]{32,64}\b`), validate: highEntropy},
		{Name: "secret_jwt", Label: "JSON Web Token", Severity: model.SevMedium,
			re: regexp.MustCompile(`\beyJ[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\b`)},
		{Name: "source_code", Label: "Proprietary source code", Severity: model.SevMedium,
			re: regexp.MustCompile(`(?m)^\s*(package |import |func |class |def |#include|public class)`)},
	}
	// Default policy: block secrets/PCI/SSN on removable + network egress; alert/audit elsewhere.
	e.policies = []Policy{
		{Classifier: "secret_aws", Channel: "*", Verdict: "block"},
		{Classifier: "secret_privkey", Channel: "*", Verdict: "block"},
		{Classifier: "pci_card", Channel: "usb", Verdict: "block"},
		{Classifier: "pci_card", Channel: "http_upload", Verdict: "block"},
		{Classifier: "pci_card", Channel: "scp", Verdict: "block"},
		{Classifier: "pci_card", Channel: "rsync", Verdict: "block"},
		{Classifier: "pii_ssn", Channel: "usb", Verdict: "block"},
		{Classifier: "pii_ssn", Channel: "*", Verdict: "alert"},
		{Classifier: "source_code", Channel: "usb", Verdict: "alert"},
		{Classifier: "source_code", Channel: "cloud", Verdict: "alert"},
		{Classifier: "*", Channel: "*", Verdict: "audit"},
	}
	return e
}

// Classifiers exposes the loaded classifiers for the API/UI.
func (e *Engine) Classifiers() []map[string]string {
	var out []map[string]string
	for _, c := range e.classifiers {
		out = append(out, map[string]string{"name": c.Name, "label": c.Label, "severity": string(c.Severity)})
	}
	return out
}

// Policies exposes the policy table for the API/UI.
func (e *Engine) Policies() []Policy { return e.policies }

// Scan classifies a text blob, returning all findings.
func (e *Engine) Scan(text string) []Finding {
	var out []Finding
	if text == "" {
		return out
	}
	for _, c := range e.classifiers {
		hits := c.re.FindAllString(text, -1)
		if len(hits) == 0 {
			continue
		}
		valid := 0
		var sample string
		for _, h := range hits {
			cand := strings.TrimSpace(h)
			if c.validate != nil && !c.validate(cand) {
				continue
			}
			valid++
			if sample == "" {
				sample = redact(cand, c.Name)
			}
		}
		if valid == 0 {
			continue
		}
		out = append(out, Finding{
			Classifier: c.Name, Label: c.Label, Severity: c.Severity,
			Matches: valid, Sample: sample,
		})
	}
	return out
}

// Verdict resolves the strongest policy verdict for a finding on a channel.
func (e *Engine) Verdict(classifier, channel string) string {
	best := ""
	rank := map[string]int{"audit": 1, "alert": 2, "block": 3}
	for _, p := range e.policies {
		if (p.Classifier == classifier || p.Classifier == "*") &&
			(p.Channel == channel || p.Channel == "*") {
			if rank[p.Verdict] > rank[best] {
				best = p.Verdict
			}
		}
	}
	if best == "" {
		best = "audit"
	}
	return best
}

func luhnValid(s string) bool {
	var digits []int
	for _, r := range s {
		if r >= '0' && r <= '9' {
			digits = append(digits, int(r-'0'))
		}
	}
	if len(digits) < 13 || len(digits) > 19 {
		return false
	}
	sum := 0
	dbl := false
	for i := len(digits) - 1; i >= 0; i-- {
		d := digits[i]
		if dbl {
			d *= 2
			if d > 9 {
				d -= 9
			}
		}
		sum += d
		dbl = !dbl
	}
	return sum%10 == 0
}

// highEntropy gates the generic-token classifier so ordinary identifiers don't trip it.
func highEntropy(s string) bool {
	return shannon(s) >= 3.5 && hasMixedClasses(s)
}

func hasMixedClasses(s string) bool {
	var hasUpper, hasLower, hasDigit bool
	for _, r := range s {
		switch {
		case r >= 'A' && r <= 'Z':
			hasUpper = true
		case r >= 'a' && r <= 'z':
			hasLower = true
		case r >= '0' && r <= '9':
			hasDigit = true
		}
	}
	return hasUpper && hasLower && hasDigit
}

func shannon(s string) float64 {
	if s == "" {
		return 0
	}
	freq := map[rune]float64{}
	for _, r := range s {
		freq[r]++
	}
	var h float64
	n := float64(len(s))
	for _, c := range freq {
		p := c / n
		h -= p * math.Log2(p)
	}
	return h
}

// redact masks a sample. PII/PAN classifiers are FULLY masked so no partial SSN/card digit is
// ever persisted to DB/console/SIEM (mirrors the agent-side policy); others keep a short hint.
// Operates on runes so multibyte input can't panic on a byte slice.
func redact(s, classifier string) string {
	r := []rune(s)
	if classifier == "pci_card" || classifier == "pii_ssn" || len(r) <= 4 {
		return strings.Repeat("*", len(r))
	}
	return string(r[:2]) + strings.Repeat("*", len(r)-4) + string(r[len(r)-2:])
}

File: agent/src/dlp.rs
//! Lightweight local DLP content inspection. Runs on the endpoint before data leaves it,
//! so sensitive content is classified at the source. The server re-inspects for defense in
//! depth, but local scanning lets the agent enforce block decisions immediately.

use regex::Regex;
use std::sync::OnceLock;

pub struct Finding {
    pub classifier: String,
    pub label: String,
    pub severity: String,
    pub matches: usize,
    pub sample: String,
}

struct Pattern {
    name: &'static str,
    label: &'static str,
    severity: &'static str,
    re: Regex,
    luhn: bool,
}

fn patterns() -> &'static Vec<Pattern> {
    static P: OnceLock<Vec<Pattern>> = OnceLock::new();
    P.get_or_init(|| {
        vec![
            Pattern {
                name: "pci_card",
                label: "Payment card number",
                severity: "high",
                re: Regex::new(r"\b(?:\d[ -]?){13,19}\b").unwrap(),
                luhn: true,
            },
            Pattern {
                name: "pii_ssn",
                label: "US Social Security Number",
                severity: "high",
                re: Regex::new(r"\b\d{3}-\d{2}-\d{4}\b").unwrap(),
                luhn: false,
            },
            Pattern {
                name: "secret_aws",
                label: "AWS access key",
                severity: "critical",
                re: Regex::new(r"\b(?:AKIA|ASIA)[0-9A-Z]{16}\b").unwrap(),
                luhn: false,
            },
            Pattern {
                name: "secret_privkey",
                label: "Private key material",
                severity: "critical",
                re: Regex::new(r"-----BEGIN (?:RSA |EC |OPENSSH |DSA |PGP )?PRIVATE KEY-----")
                    .unwrap(),
                luhn: false,
            },
        ]
    })
}

pub fn scan(text: &str) -> Vec<Finding> {
    let mut out = Vec::new();
    if text.is_empty() {
        return out;
    }
    for p in patterns() {
        let mut count = 0;
        let mut sample = String::new();
        for m in p.re.find_iter(text) {
            let s = m.as_str().trim();
            if p.luhn && !luhn_valid(s) {
                continue;
            }
            count += 1;
            if sample.is_empty() {
                // Fully mask PII/PAN at the source so no partial value is ever persisted to
                // the DB / console / SIEM. Other classifiers keep a first-2/last-2 hint.
                let full = matches!(p.name, "pci_card" | "pii_ssn");
                sample = redact(s, full);
            }
        }
        if count > 0 {
            out.push(Finding {
                classifier: p.name.into(),
                label: p.label.into(),
                severity: p.severity.into(),
                matches: count,
                sample,
            });
        }
    }
    out
}

fn luhn_valid(s: &str) -> bool {
    let digits: Vec<u32> = s.chars().filter_map(|c| c.to_digit(10)).collect();
    if digits.len() < 13 || digits.len() > 19 {
        return false;
    }
    let mut sum = 0u32;
    let mut dbl = false;
    for &d in digits.iter().rev() {
        let mut x = d;
        if dbl {
            x *= 2;
            if x > 9 {
                x -= 9;
            }
        }
        sum += x;
        dbl = !dbl;
    }
    sum % 10 == 0
}

fn redact(s: &str, full: bool) -> String {
    // Operate on chars, not bytes — byte slicing panics on multibyte UTF-8.
    let chars: Vec<char> = s.chars().collect();
    let n = chars.len();
    if full || n <= 4 {
        return "*".repeat(n);
    }
    let head: String = chars[..2].iter().collect();
    let tail: String = chars[n - 2..].iter().collect();
    format!("{}{}{}", head, "*".repeat(n - 4), tail)
}

#[cfg(test)]
mod redact_tests {
    use super::*;

    // N5: PII/PAN samples must be fully masked — no digit of the card/SSN may persist.
    #[test]
    fn pii_pan_fully_masked() {
        let f = scan("payment 4111 1111 1111 1111 ssn 123-45-6789 done");
        let card = f.iter().find(|x| x.classifier == "pci_card").expect("card");
        assert!(
            card.sample.chars().all(|c| c == '*'),
            "card sample not fully masked: {}",
            card.sample
        );
        let ssn = f.iter().find(|x| x.classifier == "pii_ssn").expect("ssn");
        assert!(
            ssn.sample.chars().all(|c| c == '*'),
            "ssn sample not fully masked: {}",
            ssn.sample
        );
        assert!(!ssn.sample.contains('6'), "ssn digit leaked: {}", ssn.sample);
    }

    // Non-PII classifiers keep the first-2/last-2 hint.
    #[test]
    fn secrets_keep_partial_hint() {
        assert_eq!(redact("AKIAIOSFODNN7EXAMPLE", false).chars().filter(|&c| c == '*').count(), 16);
    }
}
